mlflow/types/schema.py

```diff
@@ -408,3 +408,236 @@ def __eq__(self, other) -> bool:
 
     def __repr__(self) -> str:
         return repr(self.inputs)
+
+
+@experimental
+class ParamSpec:
+    """
+    Specification used to represent parameters for the model.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        type: Union[DataType, str],  # pylint: disable=redefined-builtin
```

Comment:
@serena-ruan I constructed a signature containing a param `foo` with type `long` and shape (-1,). Then, at inference time, I passed in a float list instead of a long list for `foo`. Surprisingly, no exception was thrown:

```
import mlflow
from mlflow.models import infer_signature

import numpy as np

s = infer_signature(["k"], params={"foo": [8]})


class MyMod(mlflow.pyfunc.PythonModel):
    def predict(self, ctx, inp, params):
        print("PARAMS", params)
        return [7 for item in inp]


with mlflow.start_run():
    m = mlflow.pyfunc.log_model(
        python_model=MyMod(),
        artifact_path="foo",
        signature=s,
        pip_requirements=[],
    )
    print(m.signature)

m = mlflow.pyfunc.load_model(m.model_uri)
print(m.predict(["k"], params={"foo": [6.5]}))
```

```
inputs:
  [string]
outputs:
  None
params:
  ['foo': long (default: [8]) (shape: (-1,))]

PARAMS {'foo': [6.5]}
[7]
```

Seems like we're probably missing some validation logic on the array dtype. Can we fix this and add test cases?

Reply:
@serena-ruan Looks like the problem also occurs if `foo` is a param of type array of string but I pass an array of floats at inference time:

```
import mlflow
from mlflow.models import infer_signature

import numpy as np

s = infer_signature(["k"], params={"foo": ["str"]})


class MyMod(mlflow.pyfunc.PythonModel):
    def predict(self, ctx, inp, params):
        print("PARAMS", params)
        return [7 for item in inp]


with mlflow.start_run():
    m = mlflow.pyfunc.log_model(
        python_model=MyMod(),
        artifact_path="foo",
        signature=s,
        pip_requirements=[],
    )
    print(m.signature)

m = mlflow.pyfunc.load_model(m.model_uri)
print(m.predict(["k"], params={"foo": [6.5]}))
```

```
inputs:
  [string]
outputs:
  None
params:
  ['foo': string (default: ['str']) (shape: (-1,))]

PARAMS {'foo': [6.5]}
[7]
```

We should definitely fix this and add a test case too

Reply:
@serena-ruan Looks like it's not just arrays. I tried the same thing where I defined a scalar string parameter and passed a float at inference time. No exception was thrown:

```
import mlflow
from mlflow.models import infer_signature

import numpy as np

s = infer_signature(["k"], params={"foo": "str"})


class MyMod(mlflow.pyfunc.PythonModel):
    def predict(self, ctx, inp, params):
        print("PARAMS", params)
        return [7 for item in inp]


with mlflow.start_run():
    m = mlflow.pyfunc.log_model(
        python_model=MyMod(),
        artifact_path="foo",
        signature=s,
        pip_requirements=[],
    )
    print(m.signature)

m = mlflow.pyfunc.load_model(m.model_uri)
print(m.predict(["k"], params={"foo": 6.5}))
```

```
inputs:
  [string]
outputs:
  None
params:
  ['foo': string (default: str)]

PARAMS {'foo': 6.5}
[7]
```

Looks like the value remained a float too. We should definitely throw if the param isn't of type `str`.

Reply:
Finally, the same problem occurs with a scalar long parameter and passing a scalar float value at inference time:

```
import mlflow
from mlflow.models import infer_signature

import numpy as np

s = infer_signature(["k"], params={"foo": 8})


class MyMod(mlflow.pyfunc.PythonModel):
    def predict(self, ctx, inp, params):
        print("PARAMS", params)
        return [7 for item in inp]


with mlflow.start_run():
    m = mlflow.pyfunc.log_model(
        python_model=MyMod(),
        artifact_path="foo",
        signature=s,
        pip_requirements=[],
    )
    print(m.signature)

m = mlflow.pyfunc.load_model(m.model_uri)
print(m.predict(["k"], params={"foo": 6.5}))
```

```
inputs:
  [string]
outputs:
  None
params:
  ['foo': long (default: 8)]

PARAMS {'foo': 6.5}
[7]
```

These would all be excellent test cases to write once we've fixed the behavior :)

Reply:
@serena-ruan Looks like the same problems occur when querying the MLflow Model server too:

```
import mlflow
from mlflow.models import infer_signature

import numpy as np

s = infer_signature(["k"], params={"foo": "baz"})


class MyMod(mlflow.pyfunc.PythonModel):
    def predict(self, ctx, inp, params):
        print("PARAMS", params)
        return [7 for item in inp]


with mlflow.start_run():
    m = mlflow.pyfunc.log_model(
        python_model=MyMod(),
        artifact_path="foo",
        signature=s,
        pip_requirements=[],
    )
    print(m.signature)

m = mlflow.pyfunc.load_model(m.model_uri)
print(m.predict(["k"], params={"foo": 6.5}))
```

```
# cd to model folder
mlflow models serve -m . --env-manager local --port 8000
```

```
import requests

resp = requests.post(
    url="http://127.0.0.1:8000/invocations",
    json={
        "dataframe_records": [{"test": "input"}],
        "params": {
            "foo": 8.5,
        }
    },
)

print(resp)
print(resp.text)
```

```
<Response [200]>
{"predictions": [7]}
```

Can we add test cases for model serving as well?

Seems like we might need to fix the Databricks scoring server implementation to correctly perform this validation too, right? Can we fix the Databricks PR and add test cases?

Reply:
@serena-ruan One other thing we should test and fix is defining a string / long / float array parameter and then passing a list of mixed types at inference time. This currently works but shouldn't:

```
import mlflow
from mlflow.models import infer_signature

import numpy as np

s = infer_signature(["k"], params={"foo": ["baz"]})


class MyMod(mlflow.pyfunc.PythonModel):
    def predict(self, ctx, inp, params):
        print("PARAMS", params)
        return [7 for item in inp]


with mlflow.start_run():
    m = mlflow.pyfunc.log_model(
        python_model=MyMod(),
        artifact_path="foo",
        signature=s,
        pip_requirements=[],
        registered_model_name="testmodel",
    )
    print(m.signature)

m = mlflow.pyfunc.load_model(m.model_uri)
print(m.predict(["k"], params={"foo": ["bar", 0.3]}))
```

```
inputs:
  [string]
outputs:
  None
params:
  ['foo': string (default: ['baz']) (shape: (-1,))]

PARAMS {'foo': ['bar', 0.3]}
[7]
```

Reply:
There're two things:
1. For PythonModel, we didn't implement `predict` method, so user has to call _enforce_params_schema themselves in order to constraint the params passed over.
2. We don't throw exception if the parameter you passed can be converted to the schema's type: e.g. passing a float to a string type param, we automatically convert it to string. We only throw an exception when the conversion fails, say converting "str" to an integer.

Thanks for those test cases though!! I added some new test cases and make sure they can cover most of the situations.

Reply:
```
class PythonModel:
   def predict(self, ..):
     _enforce_params_schema(...)
     self._predict(...)
   def _predict(self, ...):
      "user override...this "
```

Reply:
Option 2: Use decorator to wrap `_enforce_params_schema` logic

Reply:
@serena-ruan I don't really understand what the problem is...

```
import mlflow
from mlflow.models import infer_signature

import numpy as np

s = infer_signature(["k"], params={"foo": np.datetime64("323432"), "bar": [True, False]})


class MyMod(mlflow.pyfunc.PythonModel):
    def predict(self, ctx, inp, params):
        print("PARAMS", params)
        return [7 for item in inp]


with mlflow.start_run():
    m = mlflow.pyfunc.log_model(
        python_model=MyMod(),
        artifact_path="foo",
        signature=s,
        pip_requirements=[],
        registered_model_name="testmodel",
    )
    print(m.signature)

m = mlflow.pyfunc.load_model(m.model_uri)
print("TYPE", type(m))
print(hasattr(m, "predict"))
```

This prints

```
TYPE <class 'mlflow.pyfunc.PyFuncModel'>
True
```

So, we're wrapping the PythonModel in an instance of `mlflow.pyfunc.PyFuncModel`, which means we should be able to do schema enforcement there :).

We definitely need to cover all of the situations.

Regarding the casting of float to string, that seems okay, but we definitely shouldn't cast list[float] to list[string].

Reply:
Oh yes you're right :) We have a _PythonModelPyfuncWrapper, sorry about the confusion!

Regarding the type conversion, we can do no conversion at all? Or do you prefer type conversion for scalars and not for list?

And in order to make python native primitives types compatible with numpy types, I'm thinking of adding a new field in DataType:

def __new__(cls, value, numpy_type, spark_type, pandas_type=None, python_type=None):

So that we could better identify if the parameter passed over is compatible with the numpy_type defined in ParamSchema. This is for matching for example python's `bool` with  numpy's `np.bool_`, as they're not the same.

Reply:
No problem at all :). Even _PythonModelPyFuncWrapper gets wrapped by mlflow.pyfunc.PyFuncModel, right? So it seems like mlflow.pyfunc.PyFuncModel is the right place to validate.

Reply:
I think we should throw when the type doesn’t match, even for scalars. We should also allow int -> float and float -> double casting, but not the other way around (same behavior as for input schema type enforcement)

Adding python_type to ParamSpec makes sense to me.

Reply:
So we should allow int -> long -> float -> double; what about string and bytes, do we allow the conversion? And for these allowed conversion, do we convert both for scalar and scalar inside list?

Reply:
For these allowed conversions, we should allow for scalar and list, since input conversion allows this for list.

For string and bytes, we should do whatever input schema validation does :)